assets/js/wpjm-stats.js

```diff
@@ -2,106 +2,96 @@
-			const date = new Date();
-			const now  = date.getTime();
-			const expiration = parseInt( window.localStorage[ name ], 10 );
-			return expiration >= now;
-		}
-		return false;
-	}
-
-	function setUniques( uniquesToSet ) {
-		uniquesToSet.forEach( function( uniqueKey ) {
-			updateDailyUnique( uniqueKey );
-		} );
-	}
-
-	window.wpjmLogStats = window.wpjmLogStats || function ( statsToRecord, uniquesToSet ) {
-		const jobStatsSettings = window.job_manager_stats;
-		const ajaxUrl          = jobStatsSettings.ajax_url;
-		const ajaxNonce        = jobStatsSettings.ajax_nonce;
-
-		uniquesToSet = uniquesToSet || [];
-		statsToRecord = statsToRecord || [];
-
-		if ( statsToRecord.length < 1 ) {
-			return Promise.resolve(); // Could also be an error.
 		}
-
-		const postData = new URLSearchParams( {
-			_ajax_nonce: ajaxNonce,
-			post_id: jobStatsSettings.post_id || 0,
-			action: 'job_manager_log_stat',
-			stats: statsToRecord.join( ',' ),
-		} );
-
-		return fetch( ajaxUrl, {
-			method: 'POST',
-			credentials: 'same-origin',
-			body: postData,
-		} ).finally( function () {
-			setUniques( uniquesToSet );
-		} );
 	};
+}
 
-	function hookStatsForTrigger( statsByTrigger, triggerName ) {
-		const statsToRecord    = [];
-		const uniquesToSet     = [];
-		const stats            = statsByTrigger[triggerName] || [];
-		const events           = {};
-
-		stats.forEach( function ( statToRecord ) {
-			const statToRecordKey = statToRecord.name;
-			const uniqueKey       = statToRecord.unique_key || '';
-			const isUnique        = uniqueKey.length > 0;
-
-			if ( ! isUnique ) {
-				statsToRecord.push( statToRecordKey );
-			} else {
-				if ( false === getDailyUnique( uniqueKey ) ) {
-					uniquesToSet.push( uniqueKey );
-					statsToRecord.push( statToRecordKey );
-				}
-			}
+function observeForVisibility( jobListingContainer, jobListingElement, visibleCallback, alreadyViewedListings ) {
+	const options = {
+		root: null,
+		rootMargin: "0px",
+		threshold: 1.0,
+	};
 
-			if ( statToRecord.element && statToRecord.event ) {
-				const elemToAttach = document.querySelector( statToRecord.element );
-				if ( elemToAttach && ! events[statToRecord.element] ) {
-					elemToAttach.addEventListener( statToRecord.event, function ( e ) {
-						if ( isUnique && false !== getDailyUnique( uniqueKey ) ) {
-							return;
-						}
-
-						window.wpjmStatHooks.doAction( triggerName );
-					} );
-					events[statToRecord.element] = true;
+	const observer = new IntersectionObserver(function ( entries ) {
+		entries.forEach(function ( entry ) {
+			if ( entry.isIntersecting && entry.intersectionRatio > 0.99 ) {
+				const node = entry.target;
+				if ( 1 === node.nodeType && node.classList.contains( 'job_listing' ) ) {
+					const nodeId = findIdInClassNames( node );
+					if ( nodeId > 0 && ! alreadyViewedListings[nodeId] ) {
+						alreadyViewedListings[nodeId] = true;
+						visibleCallback( node );
+					}
 				}
+				node.classList.add( 'viewed' );
```

Comment:
Seems like we track what was viewed both via this class and in `alreadyViewedListings`, would only one of them be enough?

The class would be gone when paging back and forth or running a new search, so if we don't want to log an impression again then probably the variable is better.